### PR TITLE
Migrate to Jetpack view binding

### PR DESCRIPTION
**Pull request by ionutbedregeanu**
For #5767
Migrate from Kotlin synthetics to Jetpack view binding
Remove some unused or unnecessary imports

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/exceptions/ExceptionsListFragment.kt [hunk 1]
```diff
@@ -105,29 +107,35 @@ open class ExceptionsListFragment : BaseSettingsLikeFragment(), CoroutineScope {
         inflater: LayoutInflater,
         container: ViewGroup?,
         savedInstanceState: Bundle?
-    ): View = inflater.inflate(R.layout.fragment_exceptions_domains, container, false)
+    ): View {
+        FragmentExceptionsDomainsBinding.inflate(inflater, container, false)
+        return binding.root
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/exceptions/ExceptionsListFragment.kt:112]: Seems like the binding property is not initialized?
It will also need to be set to null in `onDestroyView`.

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.kt [hunk 2]
```diff
@@ -135,30 +128,23 @@ class BrowserFragment :
...
 
     @Suppress("LongMethod", "ComplexMethod")
-    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-        val view = inflater.inflate(R.layout.fragment_browser, container, false)
-
-        statusBar = view.findViewById(R.id.status_bar_background)
-
-        popupTint = view.findViewById(R.id.popup_tint)
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
+        binding = FragmentBrowserBinding.inflate(inflater, container, false)
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.kt:132]: The binding instance needs to be freed in `onDestroyView`.

File: app/src/main/java/org/mozilla/focus/fragment/CrashReporterFragment.kt [hunk 3]
```diff
@@ -10,19 +10,23 @@ import android.view.View
...
     var onCloseTabPressed: ((sendCrashReport: Boolean) -> Unit)? = null
+    private lateinit var binding: FragmentCrashReporterBinding
 
     override fun onCreateView(
         inflater: LayoutInflater,
         container: ViewGroup?,
         savedInstanceState: Bundle?
-    ): View? = inflater.inflate(R.layout.fragment_crash_reporter, container, false)
+    ): View {
+        binding = FragmentCrashReporterBinding.inflate(inflater, container, false)
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/fragment/CrashReporterFragment.kt:27]: The binding instance needs to be freed in `onDestroyView`.
But better yet, if we only need access to views inside `onViewCreated` then we don't need to keep around the binding instance, and the ceremony for freeing it.

We can have
```
class CrashReporterFragment : Fragment(R.layout.fragment_crash_reporter) {
    // no need for the `onCreateView` override since the ctor receiving a layout will itself call `onCreateView`
    ...
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val binding = FragmentCrashReporterBinding.bind(view)
        ...
    }
}
```

File: app/src/main/java/org/mozilla/focus/fragment/FirstrunFragment.kt [hunk 4]
```diff
@@ -39,47 +38,21 @@ class FirstrunFragment : Fragment(), View.OnClickListener {
...
-        viewPager!!.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
-            override fun onPageSelected(position: Int) {
-                Onboarding.pageDisplayed.record(Onboarding.PageDisplayedExtra(0))
-
-                viewPager!!.contentDescription = adapter.getPageAccessibilityDescription(position)
-            }
-
-            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
+        binding = FragmentFirstrunBinding.inflate(inflater, container, false)
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/fragment/FirstrunFragment.kt:42]: The binding instance needs to be freed in `onDestroyView`.

File: app/src/main/java/org/mozilla/focus/fragment/FirstrunFragment.kt [hunk 5]
```diff
@@ -97,6 +70,34 @@ class FirstrunFragment : Fragment(), View.OnClickListener {
         }
     }
 
+    @Suppress("MagicNumber")
+    private fun setPager() {
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/fragment/FirstrunFragment.kt:74]: This refactoring while not needed for the ticket is a nice touchup 👍. 
Could the method be named `setupPager`? Just `set` to me says this is a setter.

File: app/src/main/java/org/mozilla/focus/searchsuggestions/ui/SearchSuggestionsFragment.kt [hunk 6]
```diff
@@ -118,29 +119,32 @@ class SearchSuggestionsFragment : Fragment(), CoroutineScope {
         inflater: LayoutInflater,
         container: ViewGroup?,
         savedInstanceState: Bundle?
-    ): View? = inflater.inflate(R.layout.fragment_search_suggestions, container, false)
+    ): View {
+        binding = FragmentSearchSuggestionsBinding.inflate(inflater, container, false)
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/searchsuggestions/ui/SearchSuggestionsFragment.kt:123]: The binding instance needs to be freed in `onDestroyView`.

File: app/src/main/java/org/mozilla/focus/session/ui/TabSheetFragment.kt [hunk 7]
```diff
@@ -18,50 +18,44 @@ import androidx.recyclerview.widget.RecyclerView
...
-
     private var scope: CoroutineScope? = null
+    private lateinit var binding: FragmentSessionssheetBinding
 
-    private lateinit var store: BrowserStore
-
-    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-        val view = inflater.inflate(R.layout.fragment_sessionssheet, container, false)
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
+        binding = FragmentSessionssheetBinding.inflate(inflater, container, false)
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/session/ui/TabSheetFragment.kt:38]: The binding instance needs to be freed in `onDestroyView`.

File: app/src/main/java/org/mozilla/focus/fragment/UrlInputFragment.kt [hunk 8]
```diff
@@ -242,28 +242,27 @@ class UrlInputFragment :
         inflater: LayoutInflater,
         container: ViewGroup?,
         savedInstanceState: Bundle?
-    ): View? {
-        val view = inflater.inflate(R.layout.fragment_urlinput, container, false)
+    ): View {
+        binding = FragmentUrlinputBinding.inflate(inflater, container, false)
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/fragment/UrlInputFragment.kt:246]: The binding instance needs to be freed in `onDestroyView`.

File: app/src/main/java/org/mozilla/focus/autocomplete/AutocompletePreference.kt [hunk 9]
```diff
@@ -1,36 +0,0 @@
...
-import android.widget.TextView
-import androidx.preference.Preference
-import androidx.preference.PreferenceViewHolder
-import org.mozilla.focus.R
-import org.mozilla.focus.ext.settings
-
-/**
- * Autocomplete preference that will show a sub screen to configure the autocomplete behavior.
- */
-class AutocompletePreference(context: Context?, attrs: AttributeSet?) : Preference(context, attrs) {
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/autocomplete/AutocompletePreference.kt:18]: This seems to have been unused for some time.
Nice to have cleaned it though it would've been nice to be in a separate commit.
If there are more such situations in code maybe a ticket just for cleaning unused classes and resources would be useful.

**Review by Mugurell — CHANGES_REQUESTED** [review]
Looks good but could use a bit of cleaning especially around following the recommended pattern of freeing to binding instance in `onDestroyView`.
Separating now extraneous changes not needed for the ticket is not really needed but I'd say is a good practice to keep in mind for the future.

**Comment by ionutbedregeanu** [comment]
I create another ticket https://github.com/mozilla-mobile/focus-android/issues/6049 to fix those code warnings which doesn't imply a major refactoring

**Comment by ionutbedregeanu** [comment]
@sv-ohorvath I marked some UI tests with Ignore since HomeScreenRobot implies an update after this migration

File: app/src/test/java/org/mozilla/focus/searchsuggestions/SearchSuggestionsViewModelTest.kt [hunk 10]
```diff
@@ -32,7 +31,6 @@ class SearchSuggestionsViewModelTest {
     @Before
     fun setup() {
         lifecycle = LifecycleRegistry(mock(LifecycleOwner::class.java))
-        lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
```
Comment (Mugurell) [inline app/src/test/java/org/mozilla/focus/searchsuggestions/SearchSuggestionsViewModelTest.kt:35]: Good find that this isn't really needed 👍

File: app/src/main/java/org/mozilla/focus/fragment/CrashReporterFragment.kt [hunk 11]
```diff
@@ -5,35 +5,27 @@
...
-        inflater: LayoutInflater,
-        container: ViewGroup?,
-        savedInstanceState: Bundle?
-    ): View? = inflater.inflate(R.layout.fragment_crash_reporter, container, false)
-
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
-
-        view.findViewById<View>(R.id.background).background =
+        val binding = FragmentCrashReporterBinding.bind(view)
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/fragment/CrashReporterFragment.kt:21]: We are trying to bind here a View which wasn't previously created - inflated.
As such this method - `onViewCreated` won't even be called.

Nice that you've removed the almost empty `onCreateView` but if not inflating the layout from there we should use the Fragment constructor taking the layout resource as a parameter like in 
```
class CrashReporterFragment : Fragment(R.layout.fragment_crash_reporter) { .. }
```

This fragment's functionality can then be tested using `about:crashtab`.

**Review by Mugurell — COMMENTED** [review]
Small change needed in `CrashReporterFragment`, LGTM otherwise.

**Review by Mugurell — APPROVED** [review]
